packages/upload-api/test/handlers/blob.js

```diff
@@ -281,11 +275,20 @@ export const test = {
       }
 
       // Invoke `blob/add` for the third time (after invoking conclude)
-      const thirdBlobAdd = await invocation.execute(connection)
+      const thirdBlobAdd = await BlobCapabilities.add
+        .invoke({ ...task, nonce: 'third' })
```

Comment:
@vasco-santos I'm running into this issue that I don't know yeh how to address so writing here as it is related to something we talked today on the call.

- Every `blob/allocate` will have a unique task because we have [`nb.cause` field](https://github.com/w3s-project/specs/blob/main/w3-blob.md#allocate-blob). I can't recall the reason why it was introduced but I think you made a case for it.
- Consequently every `http/put` also gets unique task because it references the `blob/allocate` which is also unique.
- Each `blob/accept` is also unique and depends on unique `http/put` receipt.

All of this combined with blob/add handler works out such that we have each upload same blob even if we already have it stored, which is not what we want.

This test used to pass before changes in the PR because same exact invocation was passed three times, which after this PR would just hit the cache and return same exact receipt. With this PR I had to update to use nonce to have those invocations have different CIDs, which in turn caused http/put invocation to be different in each call and therefor we are failing to find a receipt in the last one which what test depends upon.

Reply:
I think there are two possible solutions here:

1. We should check if we have a blob already in the system and if so simply issue blob/accept instead of looking up an `http/put` receipt.
2. Do a proper receipt resolution that we talked about on the call that is invocation with some awaits in them should resolve to same receipt as invocation with awaits substituted with values.
    - 😓 Doing this is non trivial because we don't identify tasks by task CID but rather by an invocation CID which in retrospect was a mistake, but also I don't think there is a simple way to change right now.

Reply:
I was making a case to put it in a receipt, but we found and alternative. The `cause` is already there since `store/*` and I think it became a pattern, as well as would allow us to have indexes over receipts as we previously talked. Also, from billing stand point, this is needed so that we know which space to bill for space. 

`blob/allocate` should only run a second time if presigned url expired though, otherwise we should be able to find a receipt for it. But maybe we should check the allocations table instead? That way we would have the CID for the `allocate` receipt?

Reply:
> would allow us to have indexes over receipts as we previously talked.

My memory is failing me here, not sure what indexing this would relate to now.

> Also, from billing stand point, this is needed so that we know which space to bill for space.

I'm not sure that's accurate because we have `nb.space` which tells us which space to bill without having to lookup through `cause`

> `blob/allocate` should only run a second time if presigned url expired though, otherwise we should be able to find a receipt for it. 

Well that is kind of the problem, we won't find a receipt if user invoked `blob/add` a second time because invocation cid will be different and consequently every other invocation due to that `cause` field. If we remove the `cause` field than we may find receipt for the same allocation.


> But maybe we should check the allocations table instead? That way we would have the CID for the `allocate` receipt?

I have updated implementation and tests which I think behave as expected now, so probably best to discuss details inline there.

Reply:
> My memory is failing me here, not sure what indexing this would relate to now.

https://github.com/w3s-project/RFC/pull/23/files#diff-2969327d85eb81a450f0d878aa3d25d484fc3404e6497f98a1db3a5cdb2db5ddR5

This is what I mean! 

> I'm not sure that's accurate because we have nb.space which tells us which space to bill without having to lookup through cause

oh yeah, you are right!

Reply:
How does removing `cause` here sounds (in a separate PR & spec) we can introduce it back when we really need it.